seat.c

```diff
@@ -273,21 +270,170 @@ handle_keyboard_key(struct wl_listener *listener, void *data)
+	char *km1_str = xkb_keymap_get_as_string(km1, XKB_KEYMAP_FORMAT_TEXT_V1);
+	char *km2_str = xkb_keymap_get_as_string(km2, XKB_KEYMAP_FORMAT_TEXT_V1);
+	bool result = strcmp(km1_str, km2_str) == 0;
+	free(km1_str);
+	free(km2_str);
+	return result;
+}
+
+static bool
+repeat_info_match(struct cg_keyboard *a, struct wlr_keyboard *b) {
+	return a->device->keyboard->repeat_info.rate == b->repeat_info.rate &&
+	       a->device->keyboard->repeat_info.delay == b->repeat_info.delay;
+}
+
+struct cg_keyboard *
+cg_keyboard_from_seat(struct cg_seat *seat, struct wlr_input_device *device) {
+	struct cg_keyboard *keyboard = calloc(1, sizeof(struct cg_keyboard));
+	if (keyboard == NULL) {
+		wlr_log(WLR_ERROR, "Could not allocate new cg_keyboard.");
+		return NULL;
+	}
+
+	keyboard->seat = seat;
+	keyboard->device = device;
+
+	return keyboard;
+}
+
+static void
+keyboard_group_add(struct cg_keyboard *keyboard) {
+	struct cg_seat *seat = keyboard->seat;
+	struct wlr_keyboard *wlr_keyboard = keyboard->device->keyboard;
+	struct cg_keyboard_group *group;
+	wl_list_for_each(group, &seat->keyboard_groups, link) {
+		struct wlr_keyboard_group *wlr_group = group->wlr_group;
+		if (keymaps_match(wlr_keyboard->keymap, wlr_group->keyboard.keymap) &&
+		   repeat_info_match(keyboard, &wlr_group->keyboard)) {
+			wlr_log(WLR_DEBUG, "Adding keyboard %p to group %p.", (void*)keyboard,
+			        (void*)wlr_group);
+			wlr_keyboard_group_add_keyboard(wlr_group, wlr_keyboard);
+			return;
+		}
+	}
+	/* This is reached if and only if the keyboard could not be inserted into
+	 * any group */
+	struct cg_keyboard_group *cg_group =
+	    calloc(1, sizeof(struct cg_keyboard_group));
+	if (cg_group == NULL) {
+		wlr_log(WLR_ERROR, "Failed to allocate keyboard group.");
+		return;
+	}
+	cg_group->wlr_group = wlr_keyboard_group_create();
+	if (cg_group->wlr_group == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create wlr keyboard group.");
+		goto cleanup;
+	}
+
+	cg_group->wlr_group->data = cg_group;
+	wlr_keyboard_set_keymap(&cg_group->wlr_group->keyboard,
+	                        keyboard->device->keyboard->keymap);
+	wlr_log(WLR_DEBUG, "Created keyboard group %p", (void*)cg_group->wlr_group);
+
+	cg_group->keyboard =
+	    cg_keyboard_from_seat(seat, cg_group->wlr_group->input_device);
+	if (cg_group->keyboard == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create keyboard from seat.");
+		goto cleanup;
+	}
+
+	wlr_log(WLR_DEBUG, "Adding keyboard %p to group %p.", (void*)keyboard,
+	        (void*)cg_group->wlr_group);
+	wlr_keyboard_group_add_keyboard(cg_group->wlr_group, wlr_keyboard);
+	wl_list_insert(&seat->keyboard_groups, &cg_group->link);
+
+	wl_signal_add(&cg_group->wlr_group->keyboard.events.key, &cg_group->key);
+	cg_group->key.notify = handle_keyboard_group_key;
+
+	wl_signal_add(&cg_group->wlr_group->keyboard.events.modifiers,
+	              &cg_group->modifiers);
+	cg_group->modifiers.notify = handle_keyboard_group_modifiers;
```

Comment:
Style: please add an empty line before the return.

Reply:
ok